libyul/backends/evm/OptimizedEVMCodeTransform.cpp

```diff
@@ -0,0 +1,527 @@
+		_assignment.variables)
+	)
+		currentSlot = varSlot;
+}
+
+OptimizedEVMCodeTransform::OptimizedEVMCodeTransform(
+	AbstractAssembly& _assembly,
+	BuiltinContext& _builtinContext,
+	bool _useNamedLabelsForFunctions,
+	CFG const& _dfg,
+	StackLayout const& _stackLayout
+):
+m_assembly(_assembly),
+m_builtinContext(_builtinContext),
+m_useNamedLabelsForFunctions(_useNamedLabelsForFunctions),
+m_dfg(_dfg),
+m_stackLayout(_stackLayout)
+{
+}
+
+void OptimizedEVMCodeTransform::assertLayoutCompatibility(Stack const& _currentStack, Stack const& _desiredStack)
+{
+	yulAssert(_currentStack.size() == _desiredStack.size(), "");
+	for (auto&& [currentSlot, desiredSlot]: ranges::zip_view(_currentStack, _desiredStack))
+		yulAssert(holds_alternative<JunkSlot>(desiredSlot) || currentSlot == desiredSlot, "");
+}
+
+AbstractAssembly::LabelID OptimizedEVMCodeTransform::getFunctionLabel(Scope::Function const& _function)
+{
+	CFG::FunctionInfo const& functionInfo = m_dfg.functionInfo.at(&_function);
+	if (!m_functionLabels.count(&functionInfo))
+		m_functionLabels[&functionInfo] = m_useNamedLabelsForFunctions ?
+			m_assembly.namedLabel(
+				functionInfo.function.name.str(),
+				functionInfo.function.arguments.size(),
+				functionInfo.function.returns.size(),
+				{}
+			) : m_assembly.newLabelId();
+	return m_functionLabels[&functionInfo];
+}
+
+void OptimizedEVMCodeTransform::validateSlot(StackSlot const& _slot, Expression const& _expression)
+{
+	std::visit(util::GenericVisitor{
+		[&](yul::Literal const& _literal) {
+			auto* literalSlot = get_if<LiteralSlot>(&_slot);
+			yulAssert(literalSlot && valueOfLiteral(_literal) == literalSlot->value, "");
+		},
+		[&](yul::Identifier const& _identifier) {
+			auto* variableSlot = get_if<VariableSlot>(&_slot);
+			yulAssert(variableSlot && variableSlot->variable.get().name == _identifier.name, "");
+		},
+		[&](yul::FunctionCall const& _call) {
+			auto* temporarySlot = get_if<TemporarySlot>(&_slot);
+			yulAssert(temporarySlot && &temporarySlot->call.get() == &_call && temporarySlot->index == 0, "");
+		}
+	}, _expression);
+}
+
+void OptimizedEVMCodeTransform::createStackLayout(Stack _targetStack)
+{
+	static constexpr auto slotVariableName = [](StackSlot const& _slot) {
+		return std::visit(util::GenericVisitor{
+			[](VariableSlot const& _var) { return _var.variable.get().name; },
+			[](auto const&) { return YulString{}; }
+		}, _slot);
+	};
+
+	yulAssert(m_assembly.stackHeight() == static_cast<int>(m_stack.size()), "");
+	// ::createStackLayout asserts that it has successfully achieved the target layout.
+	::createStackLayout(
+		m_stack,
+		_targetStack | ranges::to<Stack>,
+		// Swap callback.
+		[&](unsigned _i)
+		{
+			yulAssert(static_cast<int>(m_stack.size()) == m_assembly.stackHeight(), "");
+			yulAssert(_i > 0, "");
+			if (_i <= 16)
+				m_assembly.appendInstruction(evmasm::swapInstruction(_i));
+			else
+			{
+				int deficit = static_cast<int>(_i - 16);
```

Comment:
```suggestion
				int deficit = static_cast<int>(_i) - 16;
```
(although I'm actually not sure which one is better)

Reply:
Best would probably be to just make this ``size_t`` in ``StackTooDeepError`` :-).